### PR TITLE
Issue 1262 consul connect

### PR DESCRIPTION
**Pull request by eikenb**
Initial Consul Connect support. Target is to allow proxies (nginx, haproxy) to route directly to services inside the mesh network. This is the basic support and reference docs. I'm still working on some examples/intros for each.

### CODE REVIEW

File: dependency/connect_leaf.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+func (d *ConnectLeafQuery) Stop() {
+	close(d.stopCh)
+}
+
+func (d *ConnectLeafQuery) CanShare() bool {
+	return false
+}
+
+func (d *ConnectLeafQuery) Type() Type {
+	return TypeConsul
```
Comment (rboyer) [inline dependency/connect_leaf.go]: Do these objects need unique Types or is it fine for them to all share `TypeConsul`?

File: template/funcs.go [hunk 2]
```diff
@@ -472,6 +473,62 @@ func servicesFunc(b *Brain, used, missing *dep.Set) func(...string) ([]*dep.Cata
 	}
 }
 
+// serviceFunc returns or accumulates health service dependencies.
```
Comment (rboyer) [inline template/funcs.go]: The comment doesn't match the name of the function.

File: template/template_test.go [hunk 3]
```diff
@@ -1606,8 +1607,87 @@ func TestTemplate_Execute(t *testing.T) {
...
+		},
+		{
+			"func_connect",
+			&NewTemplateInput{
+				Contents: `{{ range connect "webapp" }}{{ .Address }}{{ end }}`,
+			},
+			&ExecuteInput{
+				Brain: func() *Brain {
+					b := NewBrain()
+					d, err := dep.NewHealthServiceQuery("webapp")
```
Comment (rboyer) [inline template/template_test.go]:
```suggestion
					d, err := dep.NewHealthConnectQuery("webapp")
```
I suspect your setup will need to be adjusted after you actually use the new code.

File: template/template_test.go [hunk 4]
```diff
@@ -1606,8 +1607,87 @@ func TestTemplate_Execute(t *testing.T) {
...
+					})
+					return b
+				}(),
+			},
+			"1.2.3.45.6.7.8",
+			false,
+		},
 	}
 
+	//	struct {
```
Comment (rboyer) [inline template/template_test.go]: stray notes?

Comment (eikenb) [reply template/template_test.go]: Sort of... notes that I thought I'd keep. I'm always referring to the struct type at the very top when adding tests at the bottom. Annoying to flip back and forth. Thought I'd just add a commented copy of the struct where I could see it.

File: template/template_test.go (hunk 3 above)
Comment (eikenb) [reply template/template_test.go]: :man_facepalming:

Comment (eikenb) [reply template/template_test.go]: And actually no other changes are needed, aside from calling the correct method. Those tests are totally mock based using the data injected into the *Brain (thinking Neo plugging in here).. and that data was correct. So as long as the NewHealthConnectQuery itself doesn't return an error it behaves the same.

File: template/funcs.go (hunk 2 above)
Comment (eikenb) [reply template/funcs.go]: All these methods are so similar that the cut-n-paste urge is high... but that leads to crap like this. :disappointed:

File: dependency/connect_leaf.go (hunk 1 above)
Comment (eikenb) [reply dependency/connect_leaf.go]: At the moment there are only 2 types. TypeConsul and TypeVault that indicate the backend. So they should both be TypeConsul.

File: template/funcs.go (hunk 2 above)
Comment (eikenb) [reply template/funcs.go]: Fixed.

File: template/template_test.go (hunk 3 above)
Comment (eikenb) [reply template/template_test.go]: Fixed

**Pull request merged**